components/br-stream/src/router.rs

```diff
@@ -60,24 +218,397 @@ impl Router {
+                table_id,
+                cf,
+                region_id,
+                is_meta,
+                ..
+            },
+            data_file,
+        ) in self.files.into_iter()
+        {
+            let mut file_meta = data_file.generate_metadata()?;
+            let path = if is_meta {
+                Self::path_to_schema_file(file_meta.meta.min_ts)
+            } else {
+                Self::path_to_log_file(table_id, file_meta.meta.min_ts)
+            };
+            file_meta.meta.set_path(path);
+            file_meta.meta.set_cf(cf);
+            file_meta.meta.set_region_id(region_id as _);
+            result.push(file_meta)
+        }
+        Ok(result)
+    }
+}
+
+/// A opened log file with some metadata.
+struct DataFile {
+    min_ts: TimeStamp,
+    max_ts: TimeStamp,
+    resolved_ts: TimeStamp,
+    sha256: Hasher,
+    inner: File,
+    start_key: Vec<u8>,
+    end_key: Vec<u8>,
+    number_of_entries: usize,
+    local_path: PathBuf,
+}
+
+#[derive(Debug)]
+pub struct DataFileWithMeta {
+    pub meta: DataFileInfo,
+    pub local_path: PathBuf,
+}
+
+#[derive(Debug)]
+pub struct MetadataOfFlush {
+    pub files: Vec<DataFileWithMeta>,
+    pub min_resolved_ts: u64,
+}
+
+impl MetadataOfFlush {
+    fn with_capacity(cap: usize) -> Self {
+        Self {
+            files: Vec::with_capacity(cap),
+            min_resolved_ts: u64::MAX,
+        }
+    }
+
+    fn push(&mut self, file: DataFileWithMeta) {
+        let rts = file.meta.resolved_ts;
+        self.min_resolved_ts = self.min_resolved_ts.min(rts);
+        self.files.push(file);
+    }
+}
+
+impl DataFile {
+    /// create and open a logfile at the path.
+    /// Note: if a file with same name exists, would truncate it.
+    async fn new(local_path: impl AsRef<Path>) -> Result<Self> {
+        let sha256 = Hasher::new(MessageDigest::sha256())
+            .map_err(|err| Error::Other(box_err!("openssl hasher failed to init: {}", err)))?;
+        Ok(Self {
+            min_ts: TimeStamp::max(),
+            max_ts: TimeStamp::zero(),
+            resolved_ts: TimeStamp::zero(),
+            inner: File::create(local_path.as_ref()).await?,
+            sha256,
+            number_of_entries: 0,
+            start_key: vec![],
+            end_key: vec![],
+            local_path: local_path.as_ref().to_owned(),
+        })
+    }
+
+    /// Add a new KV pair to the file, returning its size.
+    async fn on_event(&mut self, mut kv: ApplyEvent) -> Result<usize> {
+        let encoded = super::endpoint::Endpoint::<EtcdStore>::encode_event(&kv.key, &kv.value);
+        let size = encoded.len();
+        let key = Key::from_encoded(std::mem::take(&mut kv.key));
+        let ts = key.decode_ts().expect("key without ts");
+        self.min_ts = self.min_ts.min(ts);
+        self.max_ts = self.max_ts.max(ts);
+        self.resolved_ts = self.resolved_ts.max(kv.region_resolved_ts.into());
+        self.inner.write_all(encoded.as_slice()).await?;
```

Comment:
do we need sync_all here 🤔

Reply:
we would sync the file before copying them to S3 at `TemporaryFiles::flush`.